module/extension/Director/src/director/run_task_pack.cpp

```diff
@@ -14,15 +14,244 @@
+                                // TODO this also influences displacement in that if we displace a parent and a child
+                                // then we need to reevaluate them differently since challenge priority doesn't work
+                                // properly anymore
+                                // TODO probably need to do a displacement pass, followed by a reevaluation pass
+
+                                // TODO if we do this there is no way to clean it up later currently?
+
+                                // const auto& provider = needed_providers[j];
+                                // auto& g              = provider->group;
+
+                                // TODO if we are a parent of the currently running task, we can skip this
+                                // TODO what if the provider changes from active provider? (it shouldn't?)
+                            }
+
+                            // Run this specific task using this specific provider
+                            run_task_on_provider(new_task, main_provider);
+                        }
+                    }
+                }
+
+                // Sort the tasks so we reevaluate them in priority order
+                std::sort(displaced_tasks.begin(), displaced_tasks.end(), [this](const auto& a, const auto& b) {
+                    return !challenge_priority(a, b);
+                });
+
+                for (const auto& t : displaced_tasks) {
+                    // Everyone we displaced needs to be reevaluated
+                    reevaluate_group(providers.at(t->requester_id)->group);
+                }
+
+                return RunLevel::OK;
+            }
+        }
+
+        // We are able to push others
+        if (run_level >= RunLevel::PUSH) {
+            // TODO Queue everywhere, setup push providers
+        }
+
+        // If we reach here, we have already queued into everything that could possibly change our state
+        return RunLevel::BLOCKED;
+    }
+
     void Director::run_task_pack(const TaskPack& pack) {
-        // TODO(@TrentHouliston) write the algorithm that does what is needed with this task pack
+
+        const auto& provider = pack.first;
+        auto& group          = provider->group;
+
+        // See if a done command was emitted
+        for (const auto& t : pack.second) {
+            if (t->type == typeid(::extension::behaviour::Done)) {
+
+                auto parent_provider = providers.at(group.active_task->requester_id);
+                auto& parent_group   = parent_provider->group;
+
+                // If it's a root provider, then we just remove the task
+                if (parent_provider->classification == Provider::Classification::ROOT) {
+                    std::cout << "Removing " << parent_group.subtasks.front() << " from "
+                              << NUClear::util::demangle(parent_group.type.name()) << std::endl;
```

Comment:
`log`?

Reply:
This shouldn't be here
I should add logs at some point though